Gems/Atom/RHI/Code/Tests/DrawPacketTests.cpp

```diff
@@ -177,246 +185,503 @@ namespace UnitTest
+
+            // Test default value
+            for (uint8_t i = 0; i < drawItemCount; ++i)
+            {
+                const RHI::DrawItem* drawItem = drawPacket->m_drawItems + i;
+                EXPECT_EQ(drawItem->m_arguments.m_type, RHI::DrawType::Indexed);
+                EXPECT_EQ(drawItem->m_arguments.m_indexed.m_instanceCount, 1);
+            }
+
+            drawPacket->SetInstanceCount(12);
+
+            for (uint8_t i = 0; i < drawItemCount; ++i)
+            {
+                const RHI::DrawItem* drawItem = drawPacket->m_drawItems + i;
+                EXPECT_EQ(drawItem->m_arguments.m_indexed.m_instanceCount, 12);
+            }
+
+            delete drawPacket;
         }
 
-        size_t tagIndex = 0;
-        for (auto& drawList : listsByTag)
+        void TestSetRootConstants()
         {
-            SortDrawList(drawList, RHI::DrawListSortType::KeyThenDepth);
+            AZ::SimpleLcgRandom random(s_randomSeed);
+
+            DrawPacketData drawPacketData(random);
+
+            RHI::DrawPacketBuilder builder;
+            RHI::DrawPacket* drawPacket = const_cast<RHI::DrawPacket*>(drawPacketData.Build(builder));
+
+            AZStd::array<uint8_t, sizeof(unsigned int) * 4> rootConstantOld;
+            EXPECT_EQ(sizeof(unsigned int) * 4, drawPacket->m_rootConstantSize);
 
-            RHI::DrawListTag tag(tagIndex);
+            for (uint8_t i = 0; i < drawPacket->m_rootConstantSize; ++i)
+            {
+                rootConstantOld[i] = drawPacket->m_rootConstants[i];
+            }
+
+            AZStd::array<uint8_t, sizeof(unsigned int) * 2> rootConstantNew = { 1, 2, 3, 4, 5, 6, 7, 8 };
+
+            // Attempt to set beyond the array
+            AZ_TEST_START_TRACE_SUPPRESSION;
+            drawPacket->SetRootConstant(9, rootConstantNew);
+            AZ_TEST_STOP_TRACE_SUPPRESSION(1);
+
+            // Nothing will be set if it triggers the assert
+            for (uint8_t i = 0; i < drawPacket->m_rootConstantSize; ++i)
+            {
+                EXPECT_EQ(rootConstantOld[i], drawPacket->m_rootConstants[i]);
+            }
 
-            RHI::DrawListView drawListView = drawListContext.GetList(tag);
-            EXPECT_EQ(drawListView.size(), drawList.size());
+            drawPacket->SetRootConstant(8, rootConstantNew);
 
-            for (size_t i = 0; i < drawList.size(); ++i)
+            // Compare the part staying the same.
+            for (uint8_t i = 0; i < drawPacket->m_rootConstantSize - 8; ++i)
             {
-                EXPECT_EQ(drawList[i], drawListView[i]);
+                EXPECT_EQ(rootConstantOld[i], drawPacket->m_rootConstants[i]);
             }
 
-            tagIndex++;
+            // Compare the part being set
+            for (uint8_t i = drawPacket->m_rootConstantSize - 8; i < drawPacket->m_rootConstantSize; ++i)
+            {
+                EXPECT_EQ(rootConstantNew[i - (drawPacket->m_rootConstantSize -8)], drawPacket->m_rootConstants[i]);
+            }
+
```

Comment:
For both this test and the SetInstanceCount test, can you add something at the end that clones the draw packet, sets the root constant/instance count on the cloned draw packet, and then verifies that the original did not change? Since that's the main use-case for these functions. It will validate that our cloned draw packets aren't referring to some shared data in memory.